feat(tracking): Add command line option to provide asset files injection

ribose-jeffreylau opened the pull request:
Fixes #4

ribose-jeffreylau committed:
feat(tracking): Add command line option to provide injected resources / asset files

ribose-jeffreylau commented:
Some sources that might help with this:

- https://blog.openreplay.com/integrating-google-analytics-with-react--a-full-guide/
- https://medium.com/geekculture/how-to-use-google-analytics-on-reactjs-in-5-minutes-7f6b43017ba9

But these articles assume some sort of standard React app, whereas our implementation is slightly different.

I suspect page view events can be emitted in function `handleHashChange`.

strogonoff commented on site/spa/index.tsx:
```diff
@@ -5,6 +5,7 @@
 
 import * as React from 'react';
 import * as ReactDOM from 'react-dom/client';
+import * as ReactGA from 'react-ga4';
```
@ribose-jeffreylau Nice, I like that the lib has no dependencies, we could use it probably.

We don’t need to send a bunch of events, just initialize and record the visit to begin with…

ribose-jeffreylau commented:
@strogonoff Getting ideas from https://github.com/paneron/site-builder/issues/7#issuecomment-2575270189, I believe the GA measuring / tracking ID can be injected into manifest.json.
I see that `useRemoteUsername()` is used in the index view in RegistryKit.  Should something similar for the tracking ID be used there?  _I.e._ creating a `useGaMeasuringId()` in RegistryKit index view.  Would it make sense to live there?

RegistryKit index view:

https://github.com/paneron/registry-kit/blob/ba24c83e1f79f6736f061aa1b0f7518280c699c5/src/views/index.tsx#L294

strogonoff commented:
> @strogonoff Getting ideas from [#7 (comment)](https://github.com/paneron/site-builder/issues/7#issuecomment-2575270189), I believe the GA measuring / tracking ID can be injected into manifest.json. I see that `useRemoteUsername()` is used in the index view in RegistryKit. Should something similar for the tracking ID be used there? _I.e._ creating a `useGaMeasuringId()` in RegistryKit index view. Would it make sense to live there?
> 
> RegistryKit index view:
> 
> https://github.com/paneron/registry-kit/blob/ba24c83e1f79f6736f061aa1b0f7518280c699c5/src/views/index.tsx#L294

I don’t like the idea of CLI option for google analytics ID.

strogonoff commented:
You can allow to supply `analyticsPlugin=token`. Or more flexibly a simple JS string to be added as a script in HTML. Since we only need to record a visit it may be enough. Document attack vector (someone who can affect CLI flag used in production can inject arbitrary JS).

ribose-jeffreylau commented:
I think supplying JS string is too flexible.  Does doing so bypass type-checking, etc.?

If using CLI option is not desirable, would passing in environmental variables or a specifically-named file containing the tracking ID be better?

strogonoff commented:
> I think supplying JS string is too flexible. Does doing so bypass type-checking, etc.?

It’s a string. Whatever snippet Google issued.

> If using CLI option is not desirable, would passing in environmental variables or a specifically-named file containing the tracking ID be better?

I don’t think having anything pertaining to “Google”, “Facebook”, “Github”, etc. is a good idea and we don’t want to support analytics logic in the core.

I see these options:

- Implement and maintain a pluggable analytics framework that would register user events and integrate with third-party providers with switchable backends (Google, Matomo, etc.)
- Just allow injecting a string

strogonoff commented:
> Document attack vector (someone who can affect CLI flag used in production can inject arbitrary JS).

Actually, whoever controls CLI flags probably also has the ability to inject any JS even without this feature.

ribose-jeffreylau commented:
@strogonoff From your comment here (https://github.com/paneron/registry-kit/pull/37#issuecomment-2577718079), it seems to suggest you would want to do the injection straight in site-builder, and not having to touch any of the RegistryKit, as well as ExtensionKit, I presume?

strogonoff commented:
https://github.com/paneron/site-builder/blob/0b516609eaf3afe0cb7086d647d0234c8c3d9864/site/spa/dist.mts

ribose-jeffreylau committed:
chore(yarn): Add clean-build-template

ribose-jeffreylau committed:
docs(README): Align typical dev workflow example with latest script changes

ribose-jeffreylau committed:
chore: Make injectedEntries work

- Copy injectedEntries into dist/
- Record all injected entries
- Write injected entries to a manifest file

ribose-jeffreylau committed:
chore: Make injectedAssetsDir work

ribose-jeffreylau committed:
chore(npm): Make `yarn clean` clean up compiled/ and dist/ also

ribose-jeffreylau committed:
docs(README): Add descriptions for --injected-*

ribose-jeffreylau committed:
fix(serve): Serve url-encoded filenames by first decoding

ribose-jeffreylau committed:
fix(serve): Add more mimetypes so images are properly served

ribose-jeffreylau committed:
chore(serve): Add onWarning

ribose-jeffreylau committed:
chore(serve): Make ^C more responsive

ribose-jeffreylau committed:
refactor(injectedResources): Split into smaller components

ribose-jeffreylau committed:
chore: Fix typing

ribose-jeffreylau committed:
fix(injectResources): Fix loading of multiple injectedEntries

ribose-jeffreylau committed:
fix(injectedAssetsDir): Only include injectedEntries when in loadAssets()

ribose-jeffreylau committed:
chore: Rename loadAsset() => _loadAssets()

ribose-jeffreylau commented:
@strogonoff Is there some sort of hook that I can attach a callback to, whenever an item class is being double-click opened?  I'm trying to see if there is a way to achieve something like the following:

<img width="814" alt="Screenshot 2025-01-17 at 01 55 14" src="https://github.com/user-attachments/assets/24bf5b3f-ac1b-4114-94f3-2b9ac056d700" />

Since right now, the location bar looks kinda opaque -- everything is `/#(:)`.

ribose-jeffreylau committed:
chore: Clean up debugging logs

ribose-jeffreylau committed:
chore: Remove unused react-ga4

Might bring it back in some other form if useful.
